### PR TITLE
Make unused-extern-crate warn-by-default

### LINKED ISSUE #42591: unused_extern_crates false positives
List of crates that can suffer from the link-only false positives:
- compiler_builtins
- panic runtimes

Allocators used to, but not after redesign.

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Apart from enabling the lint, this pull request also removes existing unused crates in the codebase, and fix some amount of false positives on crates with special purposes.

Now that all false positive issues are closed, it should be possible to make it available to wider users.

Quote:
> Now that macro modularization is implemented, this is true today! *https://github.com/rust-lang/rust/issues/30849#issuecomment-286573218*

Concerns: can break some `#[deny(warnings)]`.

Close #42591

### CODE REVIEW

**Comment by retep998** [comment]
`#[deny(warnings)]` is not a huge concern because cargo caps lints for dependencies.

**Comment by ishitatsuyuki** [comment]
Blocked by #42591

**Comment by oli-obk** [comment]
I'm on mobile right now, so I can't really check, but looking at the code it seems to me that when you import a macro that does an unused extern crate import, this lint will trigger, but you can't do anything about it except complain to the author.

The macro might unconditionally depend on a crate because it might be hard to detect from the macro args, whether the extern crate is needed.

**Comment by nikomatsakis** [comment]
@oli-obk is this ever intended to become a hard error?

it seems like, in that scenario, you could also set `#[allow]`?

**Comment by nikomatsakis** [comment]
I don't really understand why this was allow before, can somebody give me some background? @ishitatsuyuki sorry for the delay btw.

**Comment by oli-obk** [comment]
Right. I got so hung up in the lint triggering inside the macro, I forgot you can attach attributes to macro invocations.

**Comment by ishitatsuyuki** [comment]
@nikomatsakis https://github.com/rust-lang/rust/issues/30849#issuecomment-170973691

**Comment by ishitatsuyuki** [comment]
I've got a draft level fix for the blocker by using a hand crafted whitelist. I have no idea if this would work or not.

Is there anyone who can give me some hints on this? @jseyfried submitted multiple PRs about the lint, but I'm afraid he's busy on something else since his activity graph is empty recently.

**Comment by nikomatsakis** [comment]
@ishitatsuyuki sorry, what are you fixing with this whitelist?

**Comment by nikomatsakis** [comment]
(I guess some errors we see in practice?)

**Comment by nikomatsakis** [comment]
I'm not really sure what's the best path forward, but I don't particularly love the whitelist. =) Maybe we aren't yet ready to make this lint default to warn.

**Comment by ishitatsuyuki** [comment]
@nikomatsakis #42591

**Comment by nikomatsakis** [comment]
@ishitatsuyuki I see. Whitelisting the builtins crate is maybe right, but allocator crate doesn't quite seem right -- there could be other allocator crates -- maybe we should instead check whether it *is* an allocator crate.

**Comment by nikomatsakis** [comment]
We can do that via the `is_allocator(CrateNum)` function on metadata -- but I'm not sure where's the best place to call it. Are you familiar with the source of the lint, or should I go digging?

**Comment by ishitatsuyuki** [comment]
Sorry, I have absolutely no understanding of rustc internals. Can you take a look for me?

**Comment by alexcrichton** [comment]
ping @nikomatsakis, do you have thoughts on the last comment?

---

As an aside, I personally feel this lint should never be turned on by default. I've very commonly used `extern crate` to get linkage to work correctly, and there's basically no way for rustc to detect that.

**Comment by nikomatsakis** [comment]
@alexcrichton 

>  I've very commonly used extern crate to get linkage to work correctly, and there's basically no way for rustc to detect that.

Can you say a bit more?

**Comment by alexcrichton** [comment]
Sure yes. The compiler interprets `extern crate` directives in order to order arguments passed to the linker. Essentially the compiler does a topological sort of the crate graph and passes arguments in that order. This is important on Unix where linker argument order matters (it'll fail if it's in the wrong order).

Many `*-sys` crates on crates.io then link statically to their contents. In other words, the rlib is a proxy for the actual native library itself. Let's say then that you're working with a library like libgit2. The C library libgit2 depends on the C library libz. As a result we have two sys crates, one libgit2-sys and one libz-sys, and then namely libgit2-sys depends on libz-sys.

To get everything to work, libgit2-sys [contains `extern crate libz_sys`](https://github.com/alexcrichton/git2-rs/blob/d17cd1954b17401dcb58660dbc51d8e0a6cf1a0f/libgit2-sys/lib.rs#L11) but it doesn't actually use anything from the crate. If I add `#![deny(unused_extern_crates)]` then a compile error declares that it's not used.

I understand this is useful for pruning dependencies that are accidentally left in, but I don't think we're at a point yet where we should turn it on by default. As a result many sys crates are likely to just start picking up a bunch of `#[allow]` directives, which I take as a proxy for a lint that shouldn't be on by default.

**Comment by petrochenkov** [comment]
>As a result many sys crates are likely to just start picking up a bunch of #[allow] directives, which I take as a proxy for a lint that shouldn't be on by default.

`bad_style` lints are in a very similar position - they are generally useful, but usually `allow`-ed in some specific subset of crates - FFI crates, which follow foreign naming conventions. Still, they are warn-by-default and making them allow-by-default would be a worse choice. So, "bunch of #[allow] directives" doesn't always mean "shouldn't be on by default".

If `exten crate` is used purely for linkage, I'd say it would benefit all if it was marked with `allow` and a commentary explaining that something special is happening here, because it's not very obvious. In the meantime crates not belonging to the special FFI subset can benefit from `unused_extern_crates` being warn-by-default.

**Comment by retep998** [comment]
I'm definitely in favor of this lint being warn by default. If you genuinely are just pulling in an external crate for the linkage, that really needs to be documented as such and having an `allow` there is a small price to pay for having a very useful lint enabled by default for the supermajority of cases where crates are not being pulled in for linkage.

**Comment by nikomatsakis** [comment]
I think I agree with @retep998 and @petrochenkov that this is a non-obvious use (and I don't mind `#[allow]` once in a while); but I do wish we had clearer guidelines here.

**Comment by ishitatsuyuki** [comment]
I agree with the points mentioned above. I would like to ask for your opinion to build a whitelist for known link-only crates or not. This should include allocator-tagged crates and compiler_builtins.

I think this isn't necessary though because if special linkages are going to be #allow with a description, it should apply to these crates too.

**Comment by ishitatsuyuki** [comment]
r? @nikomatsakis

CI passed (the failure is due to OSX builders hanging).

cc @brson I would appreciate a crater run for determining the false positives.

This would probably require a changelog as it affects many FFI crates.

File: src/librustc_lsan/lib.rs
Comment (nikomatsakis) [inline src/librustc_lsan/lib.rs]: I take it from this that you removed the whitelist? I don't think we should be warning about allocator crates, since we *know* that they server a purpose, right? 

Before, you were asking for some tips on how to detect allocator crates. Ordinarily, if we knew the `CrateNum` of the crate in question, let's call it `cnum`, then we could do this by calling `tcx.is_allocator(cnum.as_def_id())` (where `tcx` is the "type context"). However, from what I can tell, this lint is generated during name resolution, at which point the `tcx` doesn't exist yet -- moreover I don't see an obvious way to extract the crate number. 

So I'm not sure the best way to fix this just now.

Comment (oli-obk) [inline src/librustc_lsan/lib.rs]: The `potentially_unused_imports` could be kept until normal late lints are run and then use the `tcx` to do those checks.

The crate id is available here: https://github.com/rust-lang/rust/blob/445077963c55297ef1e196a3525723090fe80b22/src/librustc_resolve/build_reduced_graph.rs#L260 and could be stored in the `ExternCrate` variant.

**Comment by ishitatsuyuki** [comment]
I'm afraid is_allocator would be removed in #42727. Can you suggest something instead?

**Comment by nikomatsakis** [comment]
@ishitatsuyuki good question! I'm not really sure what's the equivalent under the new system. Perhaps one of the routines like `has_global_allocator` or `has_default_lib_allocator`?

**Comment by aidanhs** [comment]
@ishitatsuyuki I've moved this out of waiting for crater while you decide on an alternative approach/resolve the merge conflicts :)

**Comment by nikomatsakis** [comment]
@alexcrichton any tips for @ishitatsuyuki -- how they can detect whether a crate provides an allocator?

**Comment by alexcrichton** [comment]
I don't think it's relevant any more now that `#[global_allocator]` has landed. There are still false positives due to allocators, but the discussion here seems to imply that's acceptable.

**Comment by ishitatsuyuki** [comment]
Rebased. Notify me in the case of leftover or something wrong. Thanks @alexcrichton for that prompt!

File: src/librustc_tsan/lib.rs
Comment (ishitatsuyuki) [inline src/librustc_tsan/lib.rs]: @alexcrichton I think you forgot this!

**Comment by ishitatsuyuki** [comment]
While the allocator is now fine, panic runtimes have the same problem. Should I do something with that or just let it be?

**Comment by nikomatsakis** [comment]
@alexcrichton 

> the discussion here seems to imply that's acceptable.

Hmm, I don't feel that false positives due to allocators are acceptable. I think false positives due to random link arguments seems fine, since that's something rustc knows nothing about, but allocators feel like something we ought to be able to detect, no?

**Comment by ishitatsuyuki** [comment]
@nikomatsakis the global_allocator syntax actually reference the crate itself, so the false positive is no longer an issue.

File: src/librustc_tsan/lib.rs
Comment (nikomatsakis) [inline src/librustc_tsan/lib.rs]: So what would happen if this crate were imported into some other crate? In that case, does it (implicitly) get the global-allocator defined here?

File: src/test/compile-fail/panic-runtime/auxiliary/wants-panic-runtime-abort.rs
Comment (nikomatsakis) [inline src/test/compile-fail/panic-runtime/auxiliary/wants-panic-runtime-abort.rs]: This remains something we could detect, correct?

**Comment by alexcrichton** [comment]
@nikomatsakis 

> I think false positives due to random link arguments seems fine, since that's something rustc knows nothing about, but allocators feel like something we ought to be able to detect, no?

Hm I'm not sure I understand this. To me link order is no less random than "I pulled in this crate because it has an allocator in it". The compiler can detect both situations and act accordingly?

**Comment by retep998** [comment]
@alexcrichton If you don't actually use any symbols from that ``-sys` crate then normally that would mean you don't need whatever library that `-sys` crate is linking in. In my opinion, needing a `-sys` crate for its linking effects despite not pulling in any symbols through that `-sys` crate is quite weird and should be avoided when possible. If anything I *want* people to have to explicitly allow the unused extern crate lint on that extern crate.

**Comment by alexcrichton** [comment]
@retep998 

> If you don't actually use any symbols from that -sys crate then normally that would mean you don't need whatever library that -sys crate is linking in. 

This isn't true. My example above is one that doesn't need the symbols but needs the crate linked in.

> needing a -sys crate for its linking effects despite not pulling in any symbols through that -sys crate is quite weird and should be avoided when possible

This is not a situation we can avoid. How would you propose solving this problem?

**Comment by retep998** [comment]
@alexcrichton Crates needing a `-sys` crate for its linking effects rather than the symbols it provides is rather uncommon. Most examples I see are of one `-sys` crate which depends on another `-sys` crate due to linking the former statically and thus relying on symbols from the latter in a way that is invisible to rustc. I would be perfectly happy with this being the solution for those few cases:
```Rust
// Note that even though we don't use any symbols from libz-sys, the git2 library that we are
// linking in does rely on symbols from the libz library, so we have to pull in libz-sys regardless
// so that libz will be linked in.
#[allow(unused_extern_crates)]
extern crate libz_sys;
```
For any lint there will almost always be valid cases where the lint has to be allowed, that's the whole reason Rust lets people allow lints! We can't just reject a lint being enabled by default because of the existence of a few cases where the lint will need to be allowed.

I definitely agree that we should try to fix cases where rustc *does* have the knowledge to know it is being used, such as panic runtimes or global allocators. but rustc just doesn't have any information about `-sys` crates being used for their linking effects, so an explicit allow in those cases is the right thing to do.

**Comment by ishitatsuyuki** [comment]
Let me sum up:

- for @alexcrichton's concern on libz/libgit, I think the formal way is reusing the code of build script to also link the dependency. As it requires copying source now, the `extern crate` approach may be more optimal, although it has downsides like triggering the unused lint.
- for detectable crates like panic runtimes, I'm not sure what is optimal. Maintaining a whitelist costs, and as compiler_builtins have moved out of the main tree, it seems trickier now.
Also, as global allocator has been changed to actually make a reference to the object, the issue no longer applies. I see a possibility that panic runtimes will also become a global variable similarly. In the case, the only thing that remains is compiler_builtins as above.
Anyway, I have updated #42591 content to track the current state.

**Comment by nikomatsakis** [comment]
I think @retep998 pretty much described my POV. Specifically:

> Most examples I see are of one -sys crate which depends on another -sys crate due to linking the former statically and thus relying on symbols from the latter in a way that is invisible to rustc

The "invisible" bit is the key point. Panic runtimes and allocators are things that rustc itself knows about, and I definitely think we should avoid warnings then. Lints are **so** much better if you can just blindly follow their advice. I think that anybody who is just doing "normal" Rust coding should not have to second-guess this lint (or any lint) -- and, to me, that includes code that is using a specific panic runtime or a global allocator.

**Comment by ishitatsuyuki** [comment]
> The potentially_unused_imports could be kept until normal late lints are run and then use the tcx to do those checks.
>
> The crate id is available here: https://github.com/rust-lang/rust/blob/445077963c55297ef1e196a3525723090fe80b22/src/librustc_resolve/build_reduced_graph.rs#L260 and could be stored in the ExternCrate variant.

I'm planning to use [`is_panic_runtime`](https://github.com/rust-lang/rust/blob/1685c9298685f73db4fe890c1ed27b22408aaad7/src/librustc_metadata/cstore.rs#L299) and [`is_compiler_builtins`](https://github.com/rust-lang/rust/blob/1685c9298685f73db4fe890c1ed27b22408aaad7/src/librustc_metadata/cstore.rs#L309), but I have no idea how can I delay the run of lints. @oli-obk is there a particular way to do it?

**Comment by nikomatsakis** [comment]
@ishitatsuyuki there are, on the session, two methods: `add_lint` and `add_lint_diagnostic`, but those are perhaps not what you want. They will cause the lints to be printed. I suppose what you would rather need is to store the lints in some data structure somewhere, add a regular lint pass, and then during that pass fetch the information from this data structure. For that purpose, I would imagine you would add a field to `Resolver` -- this would be populated by the code in `librustc_resolve`, and then used by the lint code. This will require some plumbing in `librustc_driver` to move the map from place to place (you can sort of trace out how `resolve.trait_map` is propagated into the resolutions structure and then into the tcx to get the idea).

**Comment by alexcrichton** [comment]
I believe @nikomatsakis is now on vacation, so 

r? @petrochenkov

**Comment by petrochenkov** [comment]
I'd probably solve this by mirroring what `UNUSED_IMPORTS` lint does exactly.
Not even mirroring, but adding the logic for extern crates into all the visitors and tables dealing with unused imports (`maybe_unused_trait_imports` table most notably).
Can't say more without reimplementing this myself.

**Comment by arielb1** [comment]
Are you still working on this @ishitatsuyuki? Just a friendly ping to make sure this isn't getting lost.

**Comment by ishitatsuyuki** [comment]
I'm pretty busy these days and I won't be able to do it by the weekend, but I already have the plan and it's on my list.

**Comment by ishitatsuyuki** [comment]
r? @petrochenkov

This is ready for review. I will rebase after resolving any issues in the review.

The commit reverting a LLVM related change is for building on my local machine, and I will remove it once this is ready to merge.

**Comment by ishitatsuyuki** [comment]
I think the tests have almost passed, but if the failure is not spurious, then it's due to my partial revert of a LLVM change. Anyway, those should go out when I rebase (after this passes review).

**Comment by ishitatsuyuki** [comment]
@alexcrichton can you review or pick someone?

**Comment by eddyb** [comment]
Pinging @nikomatsakis for the review here.

File: src/Cargo.lock
Comment (nikomatsakis) [inline src/Cargo.lock]: was it necessary to update `Cargo.lock`?

Comment (ishitatsuyuki) [inline src/Cargo.lock]: Yes, because I cleaned up orphan dependencies.

I don't know any way to make Cargo regenerate the lock without bumping unrelated dependencies, like NPM.

Comment (eddyb) [inline src/Cargo.lock]: Just edit the `Cargo.lock` by hand to make the removals. Conversely, revert the unrelated changes. Cargo should then not regenerate the lock file.

File: src/librustc_typeck/check_unused.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check_unused.rs]: Two problems here:

First, we should be checked for global allocator crates as well. Otherwise, if you have the following setup, we will warn incorrectly:

- Crate A defines a `#[global_allocator] static ALLOCATOR`.
- Crate B uses Crate A just for this side-effect.

I think this is not the recommended pattern, but it still seems *legal* to me, and I think we should capture it rather than giving a false warning.

Second, we should not be downcasting to `CrateMetadata` -- more generally, `librustc_typeck` should not directly depend on `librustc_metadata`. Rather, these calls should be going through queries. For example, the line that says `!metadata.is_panic_runtime(&tcx.dep_graph)` should be as follows:

```
!tcx.is_panic_runtime(cnum.as_def_id())
```

The problem is that, at the moment, only the `is_panic_runtime` query exists, so you'll have to add the others. This is basically a matter of copy-and-pasting the `is_panic_runtime` code you find here:

- [add it to the macro in `ty/maps.rs`](https://github.com/rust-lang/rust/blob/5431f423aac345e023815d3f10b8cb840711d756/src/librustc/ty/maps.rs#L997),
- add an [implementation in `cstore_impl.rs`](https://github.com/rust-lang/rust/blob/5431f423aac345e023815d3f10b8cb840711d756/src/librustc_metadata/cstore_impl.rs#L139).

You'll want to do that for `is_compiler_builtins` and `has_global_allocator`.

Doing things through queries is the recommend style in general, but it also means that this code will play properly with incremental compilation and so forth.

(Now that I think about it, there might *still* be a bug here, in that I'm not sure if the `has_global_allocator` flag means that it has an *immediate* global allocator, or if it covers the transitive case too, but I think it's good enough for now.)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Getting very close! Found one problem with the lint code though. Thanks for your persistence, @ishitatsuyuki !

**Comment by ishitatsuyuki** [comment]
@nikomatsakis I think I have addressed your concerns. Can you review? Thanks.

**Comment by arielb1** [comment]
@ishitatsuyuki 

Niko is on vacation until mid-August. I suppose you'll want to wait for him?

**Comment by ishitatsuyuki** [comment]
@arielb1 I'm fine to wait for one or two weeks; but well faster is better if there's someone else can review this.

File: src/librustc/ty/context.rs
Comment (petrochenkov) [inline src/librustc/ty/context.rs]: You can use `NodeSet` here, like `maybe_unused_trait_imports` does.
Crate num can be obtained from node id (`tcx.sess.cstore.extern_mod_stmt_cnum(item.id)`).
In `check_unused.rs` extern crate checking can be done in the same visitor where `check_import` lives (it will also fix the issue with reporting `pub` extern crates as unused, which seems to exist in the current implementation).

File: src/librustc_tsan/lib.rs
Comment (petrochenkov) [inline src/librustc_tsan/lib.rs]: Could you explain what is happening here?

File: src/librustc/ty/context.rs
Comment (petrochenkov) [inline src/librustc/ty/context.rs]: Could you also add the `pub extern crate` case to tests for this lint (`compile-fail/lint-unused-extern-crate.rs`)?

**Comment by petrochenkov** [comment]
LGTM, I'll r+ after rebasing and addressing my comments.

File: src/librustc/ty/context.rs
Comment (ishitatsuyuki) [inline src/librustc/ty/context.rs]: I'm not sure a visitor is needed. I cannot reproduce the `pub` extern crates case you mentioned, and it's good to do the check in a linear time if possible. I will add the test case anyway.

File: src/librustc_tsan/lib.rs
Comment (ishitatsuyuki) [inline src/librustc_tsan/lib.rs]: I did what @alexcrichton forgot to do as a part of global_allocator PR.

File: src/librustc/ty/context.rs
Comment (petrochenkov) [inline src/librustc/ty/context.rs]: >I cannot reproduce the pub extern crates case you mentioned

Ah, I see, `pub`s are filtered away in resolve already.

>I'm not sure a visitor is needed.

Okay, not a big deal. I just wanted uniformity between unused imports and extern crates.

File: src/librustc/lib.rs
Comment (kennytm) [inline src/librustc/lib.rs]:
```
error[E0432]: unresolved import `libc::size_t`
   --> src\librustc\util\common.rs:166:9
    |
166 |     use libc::size_t;
    |         ^^^^^^^^^^^^ Maybe a missing `extern crate libc;`?
error: aborting due to previous error
error: Could not compile `rustc`.
```

**Comment by ishitatsuyuki** [comment]
Okay, #[cfg] is hard.
```
error: unused extern crate
  --> src\libpanic_unwind\lib.rs:50:1
   |
50 | extern crate unwind;
   | ^^^^^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> src\libpanic_unwind\lib.rs:33:9
   |
33 | #![deny(warnings)]
   |         ^^^^^^^^
   = note: #[deny(unused_extern_crates)] implied by #[deny(warnings)]
error: aborting due to previous error
```

I also seem to found some save-analysis bug:

```
ERROR:rustc_save_analysis::dump_visitor: Mis-calculated spans for path 'SpecExtend<T, I>::spec_extend'. Found 1 spans, expected 2. Found spans:
ERROR:rustc_save_analysis::dump_visitor:     'spec_extend' in src\liballoc\vec.rs, line 1794
ERROR:rustc_save_analysis::dump_visitor:     master span: src\liballoc\vec.rs:1794:9: 1794:50: `<Vec<T> as SpecExtend<T, I>>::spec_extend`
```

**Comment by kennytm** [comment]
@ishitatsuyuki The save-analysis bug is #43796.

**Comment by ishitatsuyuki** [comment]
Can someone reapprove this?

**Comment by kennytm** [comment]
`x86_64-pc-windows-msvc` (`check-aux`) failed. Legit, but what?

```
error: unused or unknown feature
  --> src\libpanic_unwind\lib.rs:39:12
   |
39 | #![feature(panic_unwind)]
   |            ^^^^^^^^^^^^
   |
note: lint level defined here
  --> src\libpanic_unwind\lib.rs:33:9
   |
33 | #![deny(warnings)]
   |         ^^^^^^^^
   = note: #[deny(unused_features)] implied by #[deny(warnings)]
error: aborting due to previous error
error: Could not compile `panic_unwind`.
```

**Comment by ishitatsuyuki** [comment]
@kennytm being unable to test on Windows... is a great productivity killer here. ;)

r? @petrochenkov

**Comment by ishitatsuyuki** [comment]
Seems I hit the line limit. Will do this cleanly when I get to my Linux box.

**Comment by ishitatsuyuki** [comment]
A fresh breakage by #43108. Will resolve conflicts soon.

**Comment by kennytm** [comment]
unused extern crate when building stage1-std in `x86_64-pc-windows-msvc` (dist and dist-alt).

```
   Compiling panic_unwind v0.0.0 (file:///C:/projects/rust/src/libpanic_unwind)
error: unused extern crate
   --> src\libstd\lib.rs:364:1
    |
364 | extern crate unwind;
    | ^^^^^^^^^^^^^^^^^^^^
    |
```

**Comment by ishitatsuyuki** [comment]
Updated. Can I request a bors delegation?

**Comment by ishitatsuyuki** [comment]
@petrochenkov: I have `#[cfg]` gated the offending line, so I think it's fixed. CI timed out, not sure it's spurious or my fault, and anyway it's hard to analyze.

As the queue is empty, please allow me to give a try. @bors r=petrochenkov

File: src/libstd/lib.rs
Comment (kennytm) [inline src/libstd/lib.rs]: `x86_64-pc-windows-gnu`, linker error at stage0-std. Perhaps include `target_env = "gnu"`  here.

```
   Compiling panic_unwind v0.0.0 (file:///C:/projects/rust/src/libpanic_unwind)
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-sysroot\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\dllcrt2.o" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-sysroot\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-sysroot\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\x86_64-pc-windows-gnu\\release\\deps\\std-3866ebe39d96786c.0.o" "-o" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\x86_64-pc-windows-gnu\\release\\deps\\std-3866ebe39d96786c.dll" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\x86_64-pc-windows-gnu\\release\\deps\\std-3866ebe39d96786c.crate.metadata.o" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\x86_64-pc-windows-gnu\\release\\deps\\std-3866ebe39d96786c.crate.allocator.o" "-nodefaultlibs" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\x86_64-pc-windows-gnu\\release\\deps" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\release\\deps" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\native\\libbacktrace\\.libs" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\native\\jemalloc\\lib" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-std\\x86_64-pc-windows-gnu\\release\\build\\compiler_builtins-347569f235afc98e\\out" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-sysroot\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-Wl,-Bstatic" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\librand-3424f662f1c93e79.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\liballoc_system-ec232e8e6e9a059b.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\libpanic_unwind-745ef183a9066eb0.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\liblibc-b1d9bfc3f6d18aec.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\liballoc-074e85369f069bb4.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\libstd_unicode-55187640dd1f836e.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\libcore-f0e6eca3a4a54c5c.rlib" "-Wl,--no-whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.PBsfXQu94fCV\\libcompiler_builtins-79e17d7682c58a38.rlib" "-shared" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage0-sysroot\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: C:\projects\rust\build\x86_64-pc-windows-gnu\native\libbacktrace\.libs\libbacktrace.a(backtrace.o):backtrace.c:(.text$unwind+0x16): undefined reference to `_Unwind_GetIPInfo'
          C:\projects\rust\build\x86_64-pc-windows-gnu\native\libbacktrace\.libs\libbacktrace.a(backtrace.o):backtrace.c:(.text$backtrace_full+0x7a): undefined reference to `_Unwind_Backtrace'
          C:\projects\rust\build\x86_64-pc-windows-gnu\native\libbacktrace\.libs\libbacktrace.a(simple.o):simple.c:(.text$simple_unwind+0x16): undefined reference to `_Unwind_GetIPInfo'
          C:\projects\rust\build\x86_64-pc-windows-gnu\native\libbacktrace\.libs\libbacktrace.a(simple.o):simple.c:(.text$backtrace_simple+0x3c): undefined reference to `_Unwind_Backtrace'
          collect2.exe: error: ld returned 1 exit status
          
error: aborting due to previous error
error: Could not compile `std`.
```

**Comment by ollie27** [comment]
I tried compiling this on `x86_64-pc-windows-msvc`. `rustc_errors` and `rustc_driver` both need `#[cfg(unix)]` applied to `extern crate libc` and `#![cfg_attr(unix, feature(libc))]` just like `libtest`.

I also noticed `compiletest` only uses `libc` on `#[cfg(any(target_os = "macos", target_os = "ios"))]` so that will need adding so it will build the next time stage0 is updated.

**Comment by ishitatsuyuki** [comment]
I'm hitting more and more external crate warnings. Should we have a cargobomb and fix them beforehand?
cc @aturon

**Comment by ishitatsuyuki** [comment]
Finally it's done. Let's see the community's reaction to the lint.

**Comment by xftroxgpx** [comment]
While attempting to compile latest rust:

```
   Compiling bootstrap v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/bootstrap)
error: unused extern crate
   --> src/bootstrap/lib.rs:126:1
    |
126 | extern crate serde;
    | ^^^^^^^^^^^^^^^^^^^
    |
note: lint level defined here
   --> src/bootstrap/lib.rs:116:9
    |
116 | #![deny(warnings)]
    |         ^^^^^^^^
    = note: #[deny(unused_extern_crates)] implied by #[deny(warnings)]

error: aborting due to previous error

error: Could not compile `bootstrap`.
```
[link to code as it is now in master](https://github.com/rust-lang/rust/blob/master/src/bootstrap/lib.rs#L116-L126)

**Comment by ishitatsuyuki** [comment]
I think you're using local compiler for bootstrapping, but well that's indeed something worth to strip. PR coming.

**Comment by xftroxgpx** [comment]
Yes I am using config.toml `rustc = "/home/xftroxgpx/.cargo/bin/rustc"` which points to latest rust nightly.
Will it work if I don't set that key in config.toml?
Thanks!

**Comment by CryZe** [comment]
This still has some false positives in some cases unfortunately. But I just added `#[allow(...)]` on top of my extern crate declaration.

**Comment by petrochenkov** [comment]
@CryZe 
Examples of false positives would be useful.

**Comment by CryZe** [comment]
Well, idk if it even shouldn't warn in my case. I have a C API, but I need to wrap it in a binary for emscripten (cause everything else doesn't really work well atm). So I just use `extern crate my_c_api;` in the binary for emscripten. This automatically exposes the whole C API to wasm / asm.js, while without the extern crate it doesn't. I'm not sure if that should be considered a proper false positive though.

**Comment by petrochenkov** [comment]
@CryZe 
I see, this is what was discussed in https://github.com/rust-lang/rust/pull/42588#issuecomment-310411092 and below.
Uses like this are undetectable in general case and you have to add `#[allow(unused)]` (and probably a comment explaining why this `extern crate` is necessary).

**Comment by gilescope** [comment]
Could a crate that's likely to be 'implicitly used' have a flag indicating that it has 'side effects' so we avoid giving false warnings. I'm guessing the types of crates where we would get false positives know who they are and would be able to flag themselves as such.

I think its really important that we have this warning on by default because if each layer of crates has one or two more crates than they need build times will increase significantly. If we're to going to build truly _huge_ stuff with Rust we need all levels of the crate ecosystem to be trying to keep their dependencies to a minimum.

**Pull request merged**
